asterius/src/Asterius/Passes/GCSections.hs

```diff
@@ -3,92 +3,99 @@
-                      es <- deps SM.! i_staging_sym -> -- should always succeed
-                      ( es <> i_child_syms_acc,
-                        o_m_acc
-                          { staticsMap = SM.insert i_staging_sym ss (staticsMap o_m_acc)
-                          }
-                      )
-                    | Just func <- SM.lookup i_staging_sym (functionMap store_mod),
-                      es <- deps SM.! i_staging_sym -> -- should always succeed
-                      ( es <> i_child_syms_acc,
-                        o_m_acc
-                          { functionMap =
-                              SM.insert
-                                i_staging_sym
-                                func
-                                (functionMap o_m_acc)
-                          }
-                      )
-                    | verbose_err ->
-                      ( i_child_syms_acc,
-                        o_m_acc
-                          { staticsMap =
-                              SM.insert
-                                ("__asterius_barf_" <> i_staging_sym)
-                                AsteriusStatics
-                                  { staticsType = ConstBytes,
-                                    asteriusStatics =
-                                      [ Serialized $
-                                          entityName i_staging_sym <> "\0"
-                                      ]
-                                  }
-                                (staticsMap o_m_acc)
-                          }
-                      )
-                    | otherwise ->
-                      (i_child_syms_acc, o_m_acc)
-            )
-            (SS.empty, i_m)
-            i_staging_syms
-        o_staging_syms = i_child_syms `SS.difference` o_acc_syms
+        step i_staging_sym (i_child_syms_acc, o_m_acc_syms, err_syms)
+          | Just es <- i_staging_sym `SM.lookup` dep_map =
+            (es <> i_child_syms_acc, o_m_acc_syms <> SS.singleton i_staging_sym, err_syms)
+          | verbose_err =
+            (i_child_syms_acc, o_m_acc_syms, err_syms <> SS.singleton i_staging_sym)
+          | otherwise =
+            (i_child_syms_acc, o_m_acc_syms, err_syms)
+
+-- | Given the reachable symbols (statics and functions) and the exported
+-- functions for a module, garbage-collect all unreachable parts of the module.
+gcModule :: SS.SymbolSet -> SS.SymbolSet -> AsteriusModule -> AsteriusModule
+gcModule mod_syms export_funcs m =
+  AsteriusModule
+    { staticsMap = statics,
+      functionMap = functions,
+      sptMap = spt_map,
+      ffiMarshalState =
+        FFIMarshalState
+          { ffiImportDecls = ffi_imports,
+            ffiExportDecls = ffi_exports
+          }
+    }
+  where
+    statics = staticsMap m `SM.restrictKeys` mod_syms
+    functions = functionMap m `SM.restrictKeys` mod_syms
+    spt_map = sptMap m `SM.restrictKeys` SM.keysSet statics
```

Comment:
Indeed, there is guarantee that all SPT entries have data segment symbol keys, so we can use `SM.keysSet statics`. But `mod_syms` should also work and we can use that to avoid computing `SM.keysSet statics`.

Reply:
Thanks!